### PR TITLE
Switch frontend over to user list items API

### LINKED ISSUE #2464: Enable paging for userlists, userlist items, and favorites
As a user, I'd like to be able to page through userlists, userlist items, and favorites instead of retrieving all of them at once.

#### Designs and Mockups
The UI needs to be changed to support this, @abdulkdawson should it be like search in that it automatically retrieves more as you scroll down the page?
- The following still needs work with paged results: 
  - 'Add to List Dialog' needs to show all the user's lists, or have a paging UI tool
  - 'Favorite' star needs to be correctly filled in (can be done via #2423)

### PR DESCRIPTION
**Pull request by rhysyngsun**
#### Pre-Flight checklist

- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Fixes #2464 

#### What's this PR do?
Generally, this PR adds support for pagination on the lists and favorites APIs, but this required a lot in terms of specifics due to a lot of usages of `items`:

- Added `item_count` to the user list serializer
- Updated user list serializer to return the image of the first item since we can no longer look that up on the frontend

#### How should this be manually tested?

You should be able to do the following:

- add/remove items to/from lists/paths
- open the drawer for user lists
- page through lists in the drawer
- view user lists / paths
- reorder learning paths

Nothing should break

#### Screenshots (if appropriate)

@pdpinch @Ferdi @abdulkdawson :
Screenshot_2020-01-08 MIT Open Learning

### CODE REVIEW

File: static/js/components/UserListItems.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+  const [{ isFinished }] = useRequest(
+    userListItemsRequest(userListId, pageSize * page, pageSize)
+  )
+  const initialLoadDone = page !== 0 || isFinished
+
+  const { items, count } = useSelector(createUserListItemsSelector(userListId))
+
+  return initialLoadDone ? (
+    <InfiniteScroll
```
Comment (mbertrand) [inline static/js/components/UserListItems.js]: After changing the limit from 50 to 10 (to match the `max_limit` value of `DefaultPagination`), when I get to the bottom of the page, a request for the next 10 is made, but then cancelled as the next set is requested:

Screen Shot 2020-01-09 at 10 56 21 AM

If I continue to try scrolling down, further requests are made far beyond the actual number of items in the list:

Screen Shot 2020-01-09 at 10 56 51 AM

File: course_catalog/serializers.py [hunk 2]
```diff
@@ -45,15 +45,15 @@ def to_representation(self, instance):
         # Pass context on to the serializers so that they have access to the user
         context = self.context
         if isinstance(instance, Bootcamp):
-            serializer = SimpleBootcampSerializer(instance, context=context)
+            serializer = BootcampSerializer(instance, context=context)
         elif isinstance(instance, Course):
-            serializer = SimpleCourseSerializer(instance, context=context)
+            serializer = CourseSerializer(instance, context=context)
```
Comment (mbertrand) [inline course_catalog/serializers.py]: With runs now included here, getting quite a few N+1 queries.  Might be worth trying to eliminate them.
```
web_1      | [2020-01-09 19:56:43] ERROR 28 [nplusone] notifiers.py:40 - [035bea53deb6] - Potential n+1 query detected on `LearningResourceRun.learningresourceofferor_set`
web_1      | [2020-01-09 19:56:43] ERROR 28 [nplusone] notifiers.py:40 - [035bea53deb6] - Potential n+1 query detected on `LearningResourceRun.instructors`
web_1      | [2020-01-09 19:56:43] ERROR 28 [nplusone] notifiers.py:40 - [035bea53deb6] - Potential n+1 query detected on `LearningResourceRun.instructors`
web_1      | [2020-01-09 19:56:43] ERROR 28 [nplusone] notifiers.py:40 - [035bea53deb6] - Potential n+1 query detected on `LearningResourceRun.courseprice_set`
web_1      | [2020-01-09 19:56:43] ERROR 28 [nplusone] notifiers.py:40 - [035bea53deb6] - Potential n+1 query detected on `LearningResourceRun.courseprice_set`
```

File: course_catalog/serializers.py [hunk 3]
```diff
@@ -45,15 +45,15 @@ def to_representation(self, instance):
...
         elif isinstance(instance, UserList):
-            serializer = SimpleUserListSerializer(instance, context=context)
+            serializer = UserListSerializer(instance, context=context)
         elif isinstance(instance, Program):
-            serializer = SimpleProgramSerializer(instance, context=context)
+            serializer = ProgramSerializer(instance, context=context)
```
Comment (mbertrand) [inline course_catalog/serializers.py]: Getting n+1 queries now for `ProgramItem`

File: course_catalog/serializers.py [hunk 4]
```diff
@@ -516,30 +492,39 @@ class Meta:
...
+        """Return the number of items in the list"""
+        return getattr(instance, "item_count", None) or instance.items.count()
+
+    def get_image_src(self, instance):
+        """Return the user list's image or the image of the first item"""
+        if instance.image_src:
+            return instance.image_src.url
+
+        # image_src is either a URLField or a ImageField
+        list_item = instance.items.order_by("position").first()
```
Comment (mbertrand) [inline course_catalog/serializers.py]: This line is creating n+1 query warnings

File: static/js/components/UserListItems.js [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+      hasMore={isFinished && items.length < count}
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply static/js/components/UserListItems.js]: With the above change, there's a noticeable delay between hitting the bottom of the page and seeing more results, a loading indicator might be helpful if it's not much trouble to add (in this or a future PR).

File: search/serializers_test.py [hunk 6]
```diff
@@ -14,8 +14,8 @@
     ProgramFactory,
     UserListFactory,
     VideoFactory,
-    UserListCourseFactory,
-    UserListUserListFactory,
+    UserListItemFactory,
+    UserListItemFactory,
```
Comment (mbertrand) [inline search/serializers_test.py]: Dupe import

**Comment by rhysyngsun** [comment]
@mbertrand feedback addressed

File: course_catalog/serializers.py [hunk 7]
```diff
@@ -504,42 +506,69 @@ def create(self, validated_data):
...
+        if list_item:
+            image_src = list_item.item.image_src
+            return image_src if isinstance(image_src, str) else image_src.url
```
Comment (mbertrand) [inline course_catalog/serializers.py]: Might be just bad data on my instance, but this line raises an exception for one of my users:

```
web_1      |   File "./course_catalog/serializers.py", line 569, in get_image_src
web_1      |     return image_src if isinstance(image_src, str) else image_src.url
web_1      | AttributeError: 'NoneType' object has no attribute 'url'
```

**Comment by rhysyngsun** [comment]
@mbertrand feedback addressed

File: static/js/lib/queries/user_list_items.js [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+export const clearUserListItems = (userListId: number) => {
+  const key = userListItemsKey(userListId)
+  return updateEntities({
+    [`${key}.items`]: body.results.map(transformListItem(userListId)),
+    [`${key}.count`]:   body.count,
+    [`${key}.next`]:   body.next
+  })
+}
+
+export const createUserListItemsSelector = (userListId: number) => {
```
[the remark this answers is not in the document]
Comment (aliceriot) [reply static/js/lib/queries/user_list_items.js]: I just took a quick look, I think this is still the case

[the remark this answers is not in the document]
Comment (rhysyngsun) [reply static/js/lib/queries/user_list_items.js]: fixed

**Pull request merged**